src/Horus/CairoSemantics/Runner.hs

```diff
@@ -135,9 +135,11 @@ interpret = iterM exec
       .= ( ( ExistentialAss
               ( \mv ->
                   let restAss' = map (builderToAss mv . fst) restAss
-                      asAtoms = concatMap (\x -> fromMaybe [x] (unAnd x)) restAss'
+                      flatten expr = fromMaybe [expr] (unAnd expr)
+                      asAtoms = concatMap flatten restAss'
                       restExp' = map fst restExp
-                   in (a mv .|| Expr.not (Expr.and (filter (/= a mv) asAtoms)))
+                      isNotPre = (`notElem` flatten (a mv))
+                   in (a mv .|| Expr.not (Expr.and (filter isNotPre asAtoms)))
                         .=> Expr.and (restAss' ++ [Expr.not (Expr.and restExp') | not (null restExp')])
               )
            , InstructionSemanticsAssertion
```

Comment:
How about a comment explaining in some detail what is going on in this branch of the interpreter, and any unintuitive choices that are made here.

At least, it would be nice to have an explanation of how it has been made "even more convoluted."